File: OCDemo_v6.5.xx/iOSDemo/AdDemo/Tools&Manager/LaunchLoadingView.m
//
//  LaunchLoadingView.m
//  iOSDemo
//
//  Created by ltz on 2025/1/18.
//

#import "LaunchLoadingView.h"
#import "AdSDKManager.h"

@interface LaunchLoadingView()

@property (nonatomic, strong) UILabel *timerLabel;
@property (nonatomic, strong) NSTimer *timer;
@property (nonatomic, assign) NSInteger seconds;
  
// Custom timer timeout duration, typically greater than the SDK's kATSplashExtraTolerateTimeoutKey setting. Can be adjusted according to your specific requirements.
#define LaunchLoadingView_Timeout 10

@end

@implementation LaunchLoadingView

#pragma mark - Singleton

+ (instancetype)sharedInstance {
    static LaunchLoadingView *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[LaunchLoadingView alloc] initWithFrame:[UIScreen mainScreen].bounds];
    });
    return instance;
}

#pragma mark - Initialization

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        // Set background color
        self.backgroundColor = [UIColor whiteColor];
        
        // Create and configure logo ImageView
        UIImageView *imageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"logo"]];
        imageView.contentMode = UIViewContentModeScaleAspectFit;
        imageView.translatesAutoresizingMaskIntoConstraints = NO;
        
        // Add ImageView to custom view
        [self addSubview:imageView];
        
        // Create timer label
        self.timerLabel = [[UILabel alloc] init];
        self.timerLabel.textAlignment = NSTextAlignmentCenter;
        self.timerLabel.textColor = [UIColor blackColor];
        self.timerLabel.font = [UIFont systemFontOfSize:16];
        self.timerLabel.text = @"";
        self.timerLabel.translatesAutoresizingMaskIntoConstraints = NO;
        [self addSubview:self.timerLabel];
        
        // Center constraints
        NSLayoutConstraint *centerXConstraint = [NSLayoutConstraint constraintWithItem:imageView attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeCenterX multiplier:1 constant:0];
        NSLayoutConstraint *centerYConstraint = [NSLayoutConstraint constraintWithItem:imageView attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeCenterY multiplier:1 constant:0];
        
        // Timer label constraints
        NSLayoutConstraint *labelCenterXConstraint = [NSLayoutConstraint constraintWithItem:self.timerLabel attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeCenterX multiplier:1 constant:0];
        NSLayoutConstraint *labelTopConstraint = [NSLayoutConstraint constraintWithItem:self.timerLabel attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:imageView attribute:NSLayoutAttributeBottom multiplier:1 constant:20];
        
        // Activate constraints
        [NSLayoutConstraint activateConstraints:@[centerXConstraint, centerYConstraint, labelCenterXConstraint, labelTopConstraint]];
    }
    return self;
}

#pragma mark - Public Methods

- (void)show {
    // If already added to the view hierarchy, return directly to avoid duplicate addition
    if (self.superview) {
        return;
    }
    
    // Please ensure UI operations are executed on the main thread
    UIWindow *keyWindow = nil;
    
    // Use new API for iOS 13 and above
    if (@available(iOS 13.0, *)) {
        NSArray<UIScene *> *windowScenes = [[[UIApplication sharedApplication] connectedScenes] allObjects];
        for (UIWindowScene *scene in windowScenes) {
            if ([scene isKindOfClass:[UIWindowScene class]]) {
                NSArray<UIWindow *> *windows = scene.windows;
                for (UIWindow *window in windows) {
                    if (window.isKeyWindow) {
                        keyWindow = window;
                        break;
                    }
                }
            }
            if (keyWindow) break;
        }
    }
    
    // Fallback for iOS versions below 13 or if the above method didn't retrieve the window
    if (!keyWindow) {
        keyWindow = [UIApplication sharedApplication].keyWindow;
    }
    
    // Last fallback: try to get the first window
    if (!keyWindow) {
        keyWindow = [[UIApplication sharedApplication].windows firstObject];
    }
    
    if (!keyWindow) {
        NSLog(@"LaunchLoadingView: Failed to get keyWindow");
        return;
    }
    
    @try {
        // Update frame to ensure it fills the entire screen
        self.frame = keyWindow.bounds;
        
        // Add to window to ensure it's on top
        [keyWindow addSubview:self];
        [keyWindow bringSubviewToFront:self];
         
        NSLog(@"LaunchLoadingView: Successfully added to window");
    } @catch (NSException *exception) {
        // Catch exception and log it
        NSLog(@"LaunchLoadingView show exception: %@", exception);
    }
    
    [self startTimer];
}

- (void)dismiss {
    // Stop the timer first to prevent further triggers
    [self stopTimer];
    
    // Please ensure UI operations are executed on the main thread
    if (!self.superview) {
        return;
    }
    
    @try {
        // Compatibility handling: remove from superview first
        [self removeFromSuperview];
        UIWindow *currentWindow = self.window;
        if (currentWindow && [currentWindow.subviews containsObject:self]) {
            [self removeFromSuperview];
        }
        
    } @catch (NSException *exception) {
        // Catch exception, log it but don't affect program execution
        NSLog(@"LaunchLoadingView dismiss exception: %@", exception);
    }
}

- (void)startTimer {
    // Stop the previous timer first
    [self stopTimer];
    
    // Reset timer
    self.seconds = 0;
    
    // Reset timeout flag
    self.localTimerTimeout = NO;
    
    // Update UI to display initial time
    self.timerLabel.text = [NSString stringWithFormat:@"%@ 00:00 ,%@:%d",
                            kLocalizeStr(@"当前"),
                            kLocalizeStr(@"总超时时间"),
                            FirstAppOpen_Timeout];
    
    // Create NSTimer and add it to the main thread's RunLoop
    // Use scheduledTimerWithTimeInterval to create timer on main thread
    // NSRunLoopCommonModes ensures it can trigger normally during scrolling and other operations
    __weak typeof(self) weakSelf = self;
    self.timer = [NSTimer scheduledTimerWithTimeInterval:1.0
                                                 repeats:YES
                                                   block:^(NSTimer * _Nonnull timer) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (strongSelf) {
            [strongSelf updateTimer];
        }
    }];
    
    // Add timer to NSRunLoopCommonModes to ensure it runs normally during UI operations
    [[NSRunLoop mainRunLoop] addTimer:self.timer forMode:NSRunLoopCommonModes];
}

#pragma mark - Private Methods

- (void)updateTimer {
    self.seconds++;
    
    // Check if timeout has been reached
    if (self.seconds >= LaunchLoadingView_Timeout) {
        // Timeout reached, automatically remove view
        NSLog(@"LaunchLoadingView timeout reached: %ld seconds", (long)self.seconds);
        [self dismiss];
        self.localTimerTimeout = YES;
        return;
    }
    
    NSInteger minutes = self.seconds / 60;
    NSInteger remainingSeconds = self.seconds % 60;
    
    // NSTimer is already running on main thread, update UI directly
    self.timerLabel.text = [NSString stringWithFormat:@"%@ %02ld:%02ld ,%@:%d",
                            kLocalizeStr(@"当前"),
                            (long)minutes,
                            (long)remainingSeconds,
                            kLocalizeStr(@"总超时时间"),
                            FirstAppOpen_Timeout];
}

- (void)stopTimer {
    if (self.timer) {
        // Ensure timer is stopped on main thread
        if ([NSThread isMainThread]) {
            [self.timer invalidate];
            self.timer = nil;
        } else {
            dispatch_sync(dispatch_get_main_queue(), ^{
                [self.timer invalidate];
                self.timer = nil;
            });
        }
    }
}

#pragma mark - Lifecycle

- (void)willMoveToSuperview:(UIView *)newSuperview {
    [super willMoveToSuperview:newSuperview];
    
    // When the view is about to be removed from superview (newSuperview is nil), perform cleanup
    if (newSuperview == nil) {
        // Stop timer
        [self stopTimer];
    }
}

- (void)didMoveToSuperview {
    [super didMoveToSuperview];
    
    // When the view is added to a new superview, you can do some initialization here
    if (self.superview) {
        // Ensure frame is correct
        if (self.window) {
            self.frame = self.window.bounds;
        }
    }
}

- (void)dealloc {
    // Ensure timer is stopped and resources are cleaned up when the object is destroyed
    [self stopTimer];
    NSLog(@"LaunchLoadingView: dealloc called");
}

@end

File: SwiftDemo/SwiftDemo/Others/Widget/LaunchLoadingView.swift
//
//  LaunchLoadingView.swift
//  SwiftDemo
//
//  Created by ltz on 2025/8/20.
//

import UIKit

class LaunchLoadingView: UIView {
    
    // MARK: - UI Properties
    private var timerLabel: UILabel!
    private var timer: DispatchSourceTimer?
    private var seconds: Int = 0
    private var timerQueue: DispatchQueue!
    
    // MARK: - Initializers
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupUI()
    }
    
    // MARK: - Setup Methods
    private func setupUI() {
        // 创建专用的串行队列用于计时器
        timerQueue = DispatchQueue(label: "com.launchloadingview.timer", qos: .default)
        
        // 设置背景色为白色
        backgroundColor = UIColor.white
        
        // 创建并设置logo ImageView
        let imageView = UIImageView(image: UIImage(named: "logo"))
        imageView.contentMode = .scaleAspectFit
        imageView.translatesAutoresizingMaskIntoConstraints = false
        
        // 将ImageView添加到自定义视图上
        addSubview(imageView)
        
        // 创建计时器标签
        timerLabel = UILabel()
        timerLabel.textAlignment = .center
        timerLabel.textColor = UIColor.black
        timerLabel.font = UIFont.systemFont(ofSize: 16)
        timerLabel.text = ""
        timerLabel.translatesAutoresizingMaskIntoConstraints = false
        addSubview(timerLabel)
        
        // 居中约束
        let centerXConstraint = NSLayoutConstraint(item: imageView, attribute: .centerX, relatedBy: .equal, toItem: self, attribute: .centerX, multiplier: 1, constant: 0)
        let centerYConstraint = NSLayoutConstraint(item: imageView, attribute: .centerY, relatedBy: .equal, toItem: self, attribute: .centerY, multiplier: 1, constant: 0)
        
        // 计时器标签约束
        let labelCenterXConstraint = NSLayoutConstraint(item: timerLabel!, attribute: .centerX, relatedBy: .equal, toItem: self, attribute: .centerX, multiplier: 1, constant: 0)
        let labelTopConstraint = NSLayoutConstraint(item: timerLabel!, attribute: .top, relatedBy: .equal, toItem: imageView, attribute: .bottom, multiplier: 1, constant: 20) // logo下方20点
        
        // 激活约束
        NSLayoutConstraint.activate([centerXConstraint, centerYConstraint, labelCenterXConstraint, labelTopConstraint])
    }
    
    // MARK: - Public Methods
    func show() {
        ATDemoLog("%@ aaaaa", UIApplication.shared.currentKeyWindow!)
        guard let window = UIApplication.shared.currentKeyWindow else { return }
        
        frame = window.bounds // 确保填满整个屏幕
        window.addSubview(self)
    }
    
    func startTimer() {
        // 先停止之前的计时器
        stopTimer()
        
        // 重置计时器
        seconds = 0
        
        // 在主线程更新UI
        DispatchQueue.main.async {
            self.timerLabel.text = "\(kLocalizeStr("Current")) 00:00 ,\(kLocalizeStr("timeout setting")):\(FirstAppOpen_Timeout)"
        }
        
        // 创建高精度计时器
        timer = DispatchSource.makeTimerSource(queue: timerQueue)
        
        if let timer = timer {
            // 设置计时器：立即开始，每1秒触发一次，允许0.1秒的误差
            timer.schedule(deadline: .now() + 1.0, repeating: 1.0, leeway: .milliseconds(100))
            
            // 设置计时器事件处理
            timer.setEventHandler { [weak self] in
                self?.updateTimer()
            }
            
            // 启动计时器
            timer.resume()
        }
    }
    
    func dismiss() {
        // 停止计时器
        stopTimer()
        
        removeFromSuperview() // 从父视图中移除
    }
    
    // MARK: - Private Methods
    private func updateTimer() {
        seconds += 1
        let minutes = seconds / 60
        let remainingSeconds = seconds % 60
        
        // 确保UI更新在主线程执行
        DispatchQueue.main.async {
            self.timerLabel.text = "\(kLocalizeStr("Current")) \(String(format: "%02d:%02d", minutes, remainingSeconds)) ,\(kLocalizeStr("timeout setting")):\(FirstAppOpen_Timeout)"
        }
    }
    
    private func stopTimer() {
        timer?.cancel()
        timer = nil
    }
    
    deinit {
        // 确保在对象销毁时停止计时器
        stopTimer()
    }
}
